daemon/k8s_watcher.go

```diff
@@ -1559,58 +1559,75 @@ func cnpNodeStatusController(ciliumV2Store cache.Store, cnp *cilium_v2.CiliumNet
 
 	waitForEPsErr := endpointmanager.WaitForEndpointsAtPolicyRev(ctx, rev)
 
-	serverRule, fromStoreErr := getUpdatedCNPFromStore(ciliumV2Store, cnp)
-	if fromStoreErr != nil {
-		logger.WithError(fromStoreErr).Error("error getting updated CNP from store")
-		return fromStoreErr
-	}
-
-	// Make a copy since the rule is a pointer, and any of its fields
-	// which are also pointers could be modified outside of this
-	// function.
-	serverRuleCpy := serverRule.DeepCopy()
-	_, ruleCopyParseErr := serverRuleCpy.Parse()
-	if ruleCopyParseErr != nil {
-		// If we can't parse the rule then we should signalize
-		// it in the status
-		log.WithError(ruleCopyParseErr).WithField(logfields.Object, logfields.Repr(serverRuleCpy)).
-			Warn("Error parsing new CiliumNetworkPolicy rule")
-	}
+	// Number of attempts to retry updating of CNP in case that Update fails
+	// due to out-of-date resource version.
+	maxAttempts := 5
 
-	var err3 error
+	var (
+		cnpUpdateErr       error
+		updateWaitDuration = time.Duration(200) * time.Millisecond
+	)
 
-	// Update the status of whether the rule is enforced on this node.
-	// If we are unable to parse the CNP retrieved from the store,
-	// or if endpoints did not reach the desired policy revision
-	// after 30 seconds, then mark the rule as not being enforced.
-	if policyImportErr != nil {
-		// OK is false here because the policy wasn't imported into
-		// cilium on this node; since it wasn't imported, it also
-		// isn't enforced.
-		err3 = updateCNPNodeStatus(serverRuleCpy, false, false, policyImportErr, rev, serverRuleCpy.Annotations)
-	} else if ruleCopyParseErr != nil {
-		// This handles the case where the initial instance of this
-		// rule was imported into the policy repository successfully
-		// (policyImportErr == nil), but, the rule has been updated
-		// in the store soon after, and is now invalid. As such,
-		// the rule is not OK because it cannot be imported due
-		// to parsing errors, and cannot be enforced because it is
-		// not OK.
-		err3 = updateCNPNodeStatus(serverRuleCpy, false, false, ruleCopyParseErr, rev, serverRuleCpy.Annotations)
-	} else {
-		// If the deadline by the above context, then not all
-		// endpoints are enforcing the given policy, and
-		// waitForEpsErr will be non-nil.
-		err3 = updateCNPNodeStatus(serverRuleCpy, waitForEPsErr == nil, true, waitForEPsErr, rev, serverRuleCpy.Annotations)
+	for numAttempts := 0; numAttempts < maxAttempts; numAttempts++ {
+
+		serverRule, fromStoreErr := getUpdatedCNPFromStore(ciliumV2Store, cnp)
```

Comment:
Do we really need to fetch it again just because `Update()` failed?

Reply:
@tgraf yes. The update might have failed because of conflicts. Also it's not fetching from kube-apiserver it's fetching from the local cache so it's fine.

Reply:
@tgraf - @aanm 's explanation is correct.